[java] NullAway added to `selenium-api` and `selenium-support`

mk868 opened the pull request:
### Description
In this PR I add nullness analyzer to the project. At the beginning I add it to the modules:

- `selenium-api`
- `selenium-support`

To activate plugin, I added it to each occurrence of `java_export` and `java_library` in the `BUILD.bazel` files.

NullAway Configuration details:

- `-Xep:NullAway:WARN` - report problems as warnings
- `-XepOpt:NullAway:AnnotatedPackages=org.openqa.selenium` - analyze this package (including subpackages)
- `-XepOpt:NullAway:JSpecifyMode=true` - [enable support for annotations on generic types](https://github.com/uber/NullAway/wiki/JSpecify-Support)

NOTE: NullAway has a special behavior, all classes located under AnnotatedPackages are treated as annotated with `@NullMarked` - [source](https://github.com/uber/NullAway/wiki/Supported-Annotations#nullmarked-and-nullunmarked). Anyway, We need to mark these classes with `@NullMarked` annotation manually to remain compliant with [JSpecify specification](https://jspecify.dev/docs/user-guide/#what-about-unannotated-types).


I don't have much experience with Bazel, but I think that the final step is to simplify the configuration, for example, by adding NullAway plugin globally in `//java/private`.

### Motivation and Context
The JSpecify nullness annotations will give developers better exposure to potential problems with their code to avoid NullPointerExceptions.
Related issue: https://github.com/SeleniumHQ/selenium/issues/14291
To verify potential problems with the Selenium source code, we can use nullness analyzers that report code issues based on JSpecify annotations.
One of plugins that is easy to configure with Bazel is [NullAway](https://github.com/uber/NullAway/wiki/Configuration#bazel), potential alternatives described [here](https://jspecify.dev/docs/whether/#nullness-checker-support-for-jspecify).
Context: https://github.com/SeleniumHQ/selenium/pull/14372#issuecomment-2293116104

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
<!--- Provide a general summary of your changes in the Title above -->

joerg1985 commented:
Why not `-Xep:NullAway:ERROR`, otherwise the huge output of the CI might swallow the warnings.
What do others think about this?

mk868 commented:
@joerg1985
At the moment we have ~80 warnings.
I think setting the ERROR level will cause the compilation to fail

I have such steps in my mind:

1. For each Selenium artifact: (`selenium-api`, `selenium-support`, `selenium-remote-driver`, ...)
   1. Add NullAway with WARNING level
   2. Annotate all classes in the module with the proper JSpecify annotations (the number of warnings should decrease)
   3. Fix bugs in the code (the number of warnings should ideally be 0)
   4. Change NullAway level to ERROR

What do you think?

diemol commented:
What is the ideal way to proceed? Should we have the build fail if this check fails, or should the CI pipeline fail?

I tend to prefer the latter. What do you think?

mk868 commented:
Agree,

So, I'll modify PR:

- I'll set NullAway reporting level to `ERROR`
- I'll make Bazel optionally run NullAway analysis via flag/tag/argument (I need to check that)
- I'll create an extra GitHub workflow job in the `ci-java.yml` file with NullAway analysis

Does it sound good?

mk868 commented:
/describe

mk868 commented:
/review

mk868 commented:
Hi @joerg1985, @diemol 
I updated the Pull Request.
Now nullness checks are part of the workflow (NullAway set to log issues with the error level) - I added `nullaway_level` flag to not destroy regular builds during development.

mk868 commented:
Branch rebased

mk868 commented:
Hello,

I just pushed two commits, with changes:

- the `trunk' branch was merged
- NullAway updated to the `0.11.3`
- the code was reformatted

I'm still not sure if the changes related to NullAway in the GitHub pipeline that I proposed fit the Selenium workflow.
If I get any suggestions for changes then I will try to improve the code

mk868 commented:
Is there any chance to merge this?
Or should I apply some changes / split the PR into several smaller ones?

diemol commented:
Are the spotbugs failures coming from the checks you added?

mk868 commented:
To be honest, these SpotBugs errors are confusing to me because there is no clear error message  
From what I see NullAway and Spotbugs, don't have common transitive dependencies which updates could corrupt spotbugs checks.

Could you re-run GitHub Workflow to see if the problem is repeatable?

diemol commented:
Yes, those errors only happen in this PR.

mk868 commented:
Thank you for pointing this out, I found the reason

Until my change, SpotBugs analysis has been run only for classes enclosed in the `java_library`.
For example: [/java/src/org/openqa/selenium/cli/BUILD.bazel](https://github.com/SeleniumHQ/selenium/blob/trunk/java/src/org/openqa/selenium/cli/BUILD.bazel) - SpotBugs analysis is run for `*.java` files.

Files that were not included in `java_library` but in `java_export` were not subject to SpotBugs analysis.
That's because `java_export` [calls](https://github.com/SeleniumHQ/selenium/blob/trunk/java/private/export.bzl#L27) the `native.java_library`, which by default has not linting tests.
For example [/java/src/org/openqa/selenium/edge/BUILD.bazel](https://github.com/SeleniumHQ/selenium/blob/trunk/java/src/org/openqa/selenium/edge/BUILD.bazel) - no SpotBugs analysis for `*.java` files.

My changes have updated the implementation so that SpotBugs analysis is also run for files enclosed in the `java_export`.
This exposed actual flaws in the code that we were not aware of.

New bugs found by SpotBugs:
 * `//java/src/org/openqa/selenium/chromium:chromium-lib-spotbugs`
   * `H C RV: Bad attempt to compute absolute value of signed 32-bit hashcode in org.openqa.selenium.chromium.ChromiumDriver.pin(String)  At ChromiumDriver.java:[line 215]` - also linted in my IDE by [Sonar](https://rules.sonarsource.com/java/type/Bug/RSPEC-2676/)
 * `//java/src/org/openqa/selenium/devtools/v85:v85-lib-spotbugs`
   * `D UC: Useless condition: it's known that code <= 399 (0x18f) at this point  At V85Network.java:[line 140]` - also linted in my IDE, condition `if (code < 300 && code > 399)` is never true
 * `//java/src/org/openqa/selenium/firefox:firefox-lib-spotbugs`
   * `M D NP: Possible null pointer dereference in org.openqa.selenium.firefox.AddHasExtensions$1.zipDirectory(Path) due to return value of called method  Dereferenced at AddHasExtensions.java:[line 104]` - probably a false flag
 * `//java/src/org/openqa/selenium:core-lib-spotbugs`
   * `M D REC: Exception is caught when Exception is not thrown in org.openqa.selenium.net.PortProber.isFree(String, int)  At PortProber.java:[line 122]` - rather cosmetic change (replace `catch (Exception e)` with `catch (IOException | RuntimeException e)`

I think this analysis is desired.
Maybe I should activate SpotBugs for `java_export` and also address these 4 errors in a dedicated PR?
What do you think?

diemol commented:
I think a different PR for that makes sense, thank you.

mk868 commented:
I merged `trunk` into my branch, SpotBugs related issues should be fixed now.  
Can we re-run checks?

diemol commented:
Apologies for the delay in responding, @mk868. I will run the CI again and merge it if everything is working well.

diemol reviewed (changes requested):
@mk868 it looks like the new check is already making CI fail. That is good. Do you think you can fix the errors and then we can merge, please?

mk868 commented:
@diemol Currently we have ~80 NullAway errors only for the `java/src/org/openqa/selenium/BUILD.bazel`:`core` library, for the whole project it's ~1250 errors.
I counted these problems using `bazel build //java/... --//java:nullaway_level=WARN 2>&1 | grep "\[NullAway\]" -c`

I can fix these errors progressively in separate PRs (as I have done so far, 2-3 changed files per PR), and then we can merge the `trunk` to this PR from time to time to see the progress.
Will it be okay?

diemol commented:
That sounds good, thank you.
